File: src/main/java/nz/co/searchwellington/controllers/SubmissionProcessingService.java
package nz.co.searchwellington.controllers;

import java.util.Calendar;
import java.util.Date;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import javax.servlet.http.HttpServletRequest;

import nz.co.searchwellington.controllers.submission.UrlProcessor;
import nz.co.searchwellington.geocoding.osm.CachingNominatimGeocodingService;
import nz.co.searchwellington.model.Feed;
import nz.co.searchwellington.model.Geocode;
import nz.co.searchwellington.model.Image;
import nz.co.searchwellington.model.Newsitem;
import nz.co.searchwellington.model.OsmId;
import nz.co.searchwellington.model.PublishedResource;
import nz.co.searchwellington.model.Resource;
import nz.co.searchwellington.model.Tag;
import nz.co.searchwellington.model.User;
import nz.co.searchwellington.model.Website;
import nz.co.searchwellington.repositories.HandTaggingDAO;
import nz.co.searchwellington.repositories.HibernateResourceDAO;
import nz.co.searchwellington.repositories.TagDAO;
import nz.co.searchwellington.utils.UrlFilters;

import org.apache.commons.lang.StringEscapeUtils;
import org.apache.commons.lang.StringUtils;
import org.apache.log4j.Logger;
import org.joda.time.DateTime;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import uk.co.eelpieconsulting.common.geo.model.Place;

import com.google.common.base.Strings;

@Component
public class SubmissionProcessingService {

	private static final String REQUEST_TITLE_NAME = "title";
	private static final String REQUEST_DATE_NAME = "date";
    private static final String REQUEST_DESCRIPTION_NAME = "description";
    private static final String REQUEST_SELECTED_GEOCODE = "selectedGeocode";
	private static final String REQUEST_EMBARGO_DATE_NAME = "embargo_date";
    
    private Logger log = Logger.getLogger(SubmissionProcessingService.class);
    
    private CachingNominatimGeocodingService nominatimGeocodeService;
    private TagDAO tagDAO;
    private HandTaggingDAO tagVoteDAO;
	private HibernateResourceDAO resourceDAO;
	private UrlProcessor urlProcessor;
	
	@Autowired
	public SubmissionProcessingService(CachingNominatimGeocodingService nominatimGeocodingService, TagDAO tagDAO, HandTaggingDAO tagVoteDAO, HibernateResourceDAO resourceDAO, UrlProcessor urlProcessor) {
		this.nominatimGeocodeService = nominatimGeocodingService;
		this.tagDAO = tagDAO;
		this.tagVoteDAO = tagVoteDAO;
		this.resourceDAO = resourceDAO;
		this.urlProcessor = urlProcessor;
	}
	
	public void processTitle(HttpServletRequest req, Resource editResource) {           
        if (req.getParameter(REQUEST_TITLE_NAME) != null) {
            String title = new String(req.getParameter(REQUEST_TITLE_NAME));
            title = UrlFilters.trimWhiteSpace(title);
            title = UrlFilters.stripHtml(title);            
            final String flattenedTitle = UrlFilters.lowerCappedSentence(title);           
            if (!flattenedTitle.equals(title)) {
                title = flattenedTitle;   
                log.info("Flatten capitalised sentence to '" + title + "'");
            }
            editResource.setName(title);
        }
    }
	
	public void processUrl(HttpServletRequest request, Resource editResource) {		
		urlProcessor.process(request, editResource);
	}
		
	public void processImage(HttpServletRequest request, Newsitem editResource, User loggedInUser) {
		Image image = (Image) request.getAttribute("image");
    	editResource.setImage(image);
    }
	
	public Geocode processGeocode(HttpServletRequest req) {      
		log.info("Starting processing of geocode.");
		if (!Strings.isNullOrEmpty(req.getParameter(REQUEST_SELECTED_GEOCODE))) {
	    	final String selectedGeocode = new String(req.getParameter(REQUEST_SELECTED_GEOCODE).trim());
	        log.info("Found selected geocode: " + selectedGeocode);
			final OsmId osmId = new OsmId(Long.parseLong(selectedGeocode.split("/")[0]), selectedGeocode.split("/")[1]);				

			final Place resolvedPlace = nominatimGeocodeService.resolveOsmId(osmId);
			log.info("Selected geocode " + selectedGeocode + " resolved to: " + resolvedPlace);
			
			return new Geocode(resolvedPlace.getAddress(), 
					resolvedPlace.getLatLong().getLatitude(), resolvedPlace.getLatLong().getLongitude(), 
					osmId.getId(), osmId.getType());	       
		}
		return null;
	}
	
	public void processDate(HttpServletRequest request, Resource editResource) {
        editResource.setDate((Date) request.getAttribute(REQUEST_DATE_NAME));
        if (editResource.getDate() == null && editResource.getId() == 0) {
            editResource.setDate(Calendar.getInstance().getTime());
        }
    }
	
	public void processEmbargoDate(HttpServletRequest request, Resource editResource) {
		editResource.setEmbargoedUntil((Date) request.getAttribute(REQUEST_EMBARGO_DATE_NAME));
	}
	
    public void processDescription(HttpServletRequest request, Resource editResource) {
        String description = request.getParameter(REQUEST_DESCRIPTION_NAME);
        if (description != null) {
        	description = StringEscapeUtils.unescapeHtml(description);
        	description = UrlFilters.stripHtml(description);
        }
        editResource.setDescription(description);
    }
    
    public void processHeld (HttpServletRequest request, Resource editResource) {
    	if (request.getParameter("has_held") != null) {
    		if (request.getParameter("held") != null) {
    			editResource.setHeld(true);
    			return;
    		}
    		editResource.setHeld(false);
    	}
    	return;
    }
        
	public void processTags(HttpServletRequest request, Resource editResource, User user) {
		log.info("Processing tags");
    	if (request.getParameter("has_tag_select") != null) {
    		processTagSelect(request, editResource, user);    		
    	}
        if (request.getParameter("additional_tags") != null) {
            processAdditionalTags(request, editResource, user);                   
        } else {
        	log.debug("No additional tag string found.");
        }    
    }
	
	@SuppressWarnings("unchecked")
	private void processTagSelect(HttpServletRequest request, Resource editResource, User user) {
		log.info("Processing tag select");
		if (request.getAttribute("tags") != null) {        	
			List<Tag> requestTagsList = (List <Tag>) request.getAttribute("tags");
			log.debug("Found tags on request: " + requestTagsList);
			Set<Tag> tags = new HashSet<Tag>(requestTagsList);
			log.info("Found " + tags.size() + " tags on the request");			
			tagVoteDAO.setUsersTagVotesForResource(editResource, user, tags);			
			
		} else {
			log.info("No tags request attribute seen; clearing users tag votes");
			tagVoteDAO.setUsersTagVotesForResource(editResource, user, new HashSet<Tag>());
		}
	}
    
	public void processPublisher(HttpServletRequest request, Resource editResource) {
    	boolean isPublishedResource = editResource instanceof PublishedResource;
    	if (isPublishedResource) {
    		if (request.getParameter("publisherName") != null && !request.getParameter("publisherName").equals("")) {
    			final String publisherName = request.getParameter("publisherName");
    			Website publisher = (Website) resourceDAO.getPublisherByName(publisherName);
    			if (publisher != null) {
    				log.info("Found publisher: " + publisher.getName());
    				((PublishedResource) editResource).setPublisher(publisher);
    			}
    		}
    	}
    }
	
	public void processAcceptance(HttpServletRequest request, Resource editResource, User loggedInUser) {
		if (editResource instanceof Newsitem) {
			if (request.getParameter("acceptedFromFeed") != null && !request.getParameter("acceptedFromFeed").equals("")) {			
				final String acceptedFromFeedUrlWords = request.getParameter("acceptedFromFeed");
				log.info("Item was accepted from a feed with url words: " + acceptedFromFeedUrlWords);
				Feed feed = resourceDAO.loadFeedByUrlWords(acceptedFromFeedUrlWords);
				if (feed != null) {
					log.info("Setting accepted from feed to: " + feed.getName());
					((Newsitem) editResource).setFeed(feed);
					((Newsitem) editResource).setAcceptedBy(loggedInUser);
					((Newsitem) editResource).setAccepted(new DateTime().toDate());	// TODO not test covered
				}
			}
		}
	}
	    
    private void processAdditionalTags(HttpServletRequest request, Resource editResource, User user) {
        String additionalTagString = request.getParameter("additional_tags").trim();
        log.debug("Found additional tag string: " + additionalTagString);
        String[] fields = additionalTagString.split(",");
        if (fields.length > 0) {
            
            for (int i = 0; i < fields.length; i++) {
                String field = fields[i].trim();
                    
                String displayName = field;
                field = cleanTagName(field);
                log.debug("Wants additional tag: " + field);
                                
                if (isValidTagName(field)) {
                    Tag existingTag = tagDAO.loadTagByName(field);
                    if (existingTag == null) {
                        log.debug("Tag '" + field + "' is a new tag. Needs to be created.");                                
                      
                        Tag newTag = tagDAO.createNewTag();                                                
                        newTag.setName(field);                       
                        newTag.setDisplayName(displayName);                        
                        tagDAO.saveTag(newTag);
                        tagVoteDAO.addTag(user, newTag, editResource);
                        
                    } else {
                        log.debug("Found an existing tag in the additional list: " + existingTag.getName() + "; adding.");
                        tagVoteDAO.addTag(user, existingTag, editResource);
                                                     
                    }
                } else {
                    log.debug("Ignoring invalid tag name: " + field);
                }
            }                    
        }
    }
    
    private boolean isValidTagName(String field) {
    	return field != null && field.length() > 0 && field.matches("[a-zA-Z0-9]*");
    }
    
    private String cleanTagName(String field) {
    	field = StringUtils.strip(field);
        field = StringUtils.remove(field, " ");    
        return field.toLowerCase().trim();
    }
    
}

File: src/main/scala/nz/co/searchwellington/controllers/SubmissionProcessingService.scala
package nz.co.searchwellington.controllers

import com.google.common.base.Strings
import jakarta.servlet.http.HttpServletRequest
import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.feeds.PlaceToGeocodeMapper
import nz.co.searchwellington.geocoding.osm.OsmIdParser
import nz.co.searchwellington.model._
import nz.co.searchwellington.model.geo.Geocode
import nz.co.searchwellington.repositories.mongo.MongoRepository
import nz.co.searchwellington.repositories.{HandTaggingDAO, TagDAO}
import nz.co.searchwellington.utils.{StringWrangling, UrlFilters}
import org.apache.commons.lang3.StringUtils
import org.apache.commons.logging.LogFactory
import org.apache.commons.text.StringEscapeUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import uk.co.eelpieconsulting.common.geo.model.OsmId

import scala.jdk.CollectionConverters.CollectionHasAsScala
@Component class SubmissionProcessingService @Autowired()(tagDAO: TagDAO,
                                                          tagVoteDAO: HandTaggingDAO,
                                                          osmIdParser: OsmIdParser, placeToGeocodeMapper: PlaceToGeocodeMapper,
                                                          mongoRepository: MongoRepository) extends ReasonableWaits with StringWrangling {

  private val log = LogFactory.getLog(classOf[SubmissionProcessingService])

  private val REQUEST_TITLE_NAME = "title"
  private val REQUEST_DATE_NAME = "date"
  private val REQUEST_DESCRIPTION_NAME = "description"
  private val REQUEST_SELECTED_GEOCODE = "selectedGeocode"
  private val REQUEST_EMBARGO_DATE_NAME = "embargo_date"

  def processImage(request: HttpServletRequest, editResource: Newsitem, loggedInUser: User): Unit = {
    val image = request.getAttribute("image").asInstanceOf[Image]
    // editResource.setImage(image)
  }

  def processGeocode(request: HttpServletRequest): Geocode = {  // TODO make an Option
    if (!Strings.isNullOrEmpty(request.getParameter(REQUEST_SELECTED_GEOCODE))) {

      val osmIdString: String = new String(request.getParameter(REQUEST_SELECTED_GEOCODE).trim)
      val osmId: OsmId = osmIdParser.parseOsmId(osmIdString)
      //val resolvedPlace: Place = nominatimGeocodeService.resolveOsmId(osmId)

     // log.info("Selected geocode " + osmIdString + " resolved to: " + resolvedPlace)

      /*
      Option(resolvedPlace).map { p =>
        placeToGeocodeMapper.mapPlaceToGeocode(p)

      }.getOrElse {
        log.warn("Could not resolve OSM id: " + osmId)
        null
      }
       */
      null

    } else {
      null
    }
  }

  def processEmbargoDate(request: HttpServletRequest, editResource: Resource): Unit = {
    //editResource.setEmbargoedUntil(request.getAttribute(REQUEST_EMBARGO_DATE_NAME).asInstanceOf[Date])
  }

  def processDescription(request: HttpServletRequest, editResource: Resource): Unit = {
    var description = request.getParameter(REQUEST_DESCRIPTION_NAME)
    if (description != null) {
      description = StringEscapeUtils.unescapeHtml4(description)
      description = UrlFilters.stripHtml(description)
    }
    //editResource.setDescription(description)
  }

  def processHeld(request: HttpServletRequest, editResource: Resource): Unit = {
    if (request.getParameter("has_held") != null) {
      if (request.getParameter("held") != null) {
      //  editResource.setHeld(true)
      } else {
       // editResource.setHeld(false)
      }
    }
  }

  def processTags(request: HttpServletRequest, editResource: Resource, user: User): Unit = {
    log.info("Processing tags")
    if (request.getParameter("has_tag_select") != null) {
      processTagSelect(request, editResource, user)
    }
    else {
      log.debug("No additional tag string found.")
    }
  }

  @SuppressWarnings(Array("unchecked"))
  private def processTagSelect(request: HttpServletRequest, editResource: Resource, user: User): Unit = {
    log.info("Processing tag select")
    if (request.getAttribute("tags") != null) {
      val requestTagsList = request.getAttribute("tags").asInstanceOf[java.util.List[Tag]].asScala.toSeq
      log.debug("Found tags on request: " + requestTagsList)
      log.info("Found " + requestTagsList.size + " tags on the request")
      //tagVoteDAO.setUsersTagVotesForResource(editResource, user, requestTagsList.toSet)

    } else {
      log.info("No tags request attribute seen; clearing users tag votes")
      //tagVoteDAO.setUsersTagVotesForResource(editResource, user, Set())
    }
  }

  def processPublisher(request: HttpServletRequest, editResource: Resource): Unit = {
    val isPublishedResource: Boolean = editResource.isInstanceOf[PublishedResource]
    if (isPublishedResource) {
      if (request.getParameter("publisherName") != null && !(request.getParameter("publisherName") == "")) {
        val publisherName: String = request.getParameter("publisherName")
        // val publisher: Website = resourceDAO.getPublisherByName(publisherName).asInstanceOf[Website]
        //if (publisher != null) {
         // log.info("Found publisher: " + publisher.title)
          // (editResource.asInstanceOf[PublishedResource]).setPublisher(publisher)
       // }
      }
    }
  }

  private def isValidTagName(field: String): Boolean = {
    field != null && field.nonEmpty && field.matches("[a-zA-Z0-9]*")
  }

  private def cleanTagName(i: String): String = {
    var cleaned = StringUtils.strip(i)
    cleaned = StringUtils.remove(cleaned, " ")
    cleaned.toLowerCase.trim
  }

}
